pkg/cel/compatibility.go

```diff
@@ -0,0 +1,269 @@
+			return false, fmt.Errorf("map parameter count mismatch: got %d, expected %d", len(outputParams), len(expectedParams))
+		}
+		return true, nil
+	}
+
+	// Check key type compatibility
+	compatible, err := AreTypesStructurallyCompatible(outputParams[0], expectedParams[0], provider)
+	if !compatible {
+		return false, fmt.Errorf("map key type incompatible: %w", err)
+	}
+
+	// Check value type compatibility
+	compatible, err = AreTypesStructurallyCompatible(outputParams[1], expectedParams[1], provider)
+	if !compatible {
+		return false, fmt.Errorf("map value type incompatible: %w", err)
+	}
+	return true, nil
+}
+
+// areStructTypesCompatible checks if struct types are structurally compatible
+// by introspecting their fields using the DeclTypeProvider.
+func areStructTypesCompatible(output, expected *cel.Type, provider *DeclTypeProvider) (bool, error) {
+	if provider == nil {
+		// Without provider, we can't introspect fields - fall back to kind check only
+		return true, nil
+	}
+
+	// Resolve DeclTypes by walking through nested type paths
+	expectedDecl := resolveDeclTypeFromPath(expected.String(), provider)
+	outputDecl := resolveDeclTypeFromPath(output.String(), provider)
+
+	// If we can't resolve both types, we can't do structural comparison
+	// Fall back to accepting it (permissive - could make this stricter)
+	if expectedDecl == nil || outputDecl == nil {
+		return true, nil
+	}
+
+	// Check that output has all required fields of expected
+	return areStructFieldsCompatible(outputDecl, expectedDecl, provider)
+}
+
+// resolveDeclTypeFromPath resolves a DeclType by walking through a nested path.
+// For example, "__type_ingressroute.spec.routes.@idx.middlewares" would:
+// 1. Strip TypeNamePrefix and look up "ingressroute" in the provider
+// 2. Find the "spec" field
+// 3. Find the "routes" field
+// 4. Get the list element type (@idx)
+// 5. Find the "middlewares" field
+func resolveDeclTypeFromPath(typePath string, provider *DeclTypeProvider) *apiservercel.DeclType {
+	if provider == nil || typePath == "" {
+		return nil
+	}
+
+	// Split the path into segments
+	segments := strings.Split(typePath, ".")
+	if len(segments) == 0 {
+		return nil
+	}
+
+	// Get the root name - keep it as-is (with or without prefix)
+	rootName := segments[0]
+
+	// Look up the root type in the provider
+	// Try first with the name as-is, then try without prefix if it has one
+	currentDecl, found := provider.FindDeclType(rootName)
+	if !found && strings.HasPrefix(rootName, TypeNamePrefix) {
+		// Try without prefix for backwards compatibility
+		shortName := strings.TrimPrefix(rootName, TypeNamePrefix)
+		currentDecl, found = provider.FindDeclType(shortName)
+	}
+	if !found {
+		return nil
+	}
+
+	// Walk through remaining path segments
+	for i := 1; i < len(segments); i++ {
+		segment := segments[i]
+
+		// Handle list element type (@idx) and map value type (@elem)
+		if segment == "@idx" || segment == "@elem" {
```

Comment:
I believe this opinionates (rightfully so) on the decl type provider, so I believe it needs a bit of comment as this doesnt come from cel